### PR TITLE
Adding graceful shutdown to server

### LINKED ISSUE #1976: Graceful shutdown of ServerFuture
**Is your feature request related to a problem? Please describe.**
When running a DNS server (via `ServerFuture`) as part of a larger application, it's desirable to include the shutdown of the DNS server as part of the shutdown workflow for the process.

**Describe the solution you'd like**
Ideally, the DNS server would support hooks to initiate a graceful shutdown, as well as notifying the application when the server shutdown has completed. During a graceful shutdown, the server would finish serving existing requests, but reject new requests.

**Describe alternatives you've considered**
NA

**Additional context**
This came up in an Istio PR, which adds a Trust-DNS-based proxy: https://github.com/istio/ztunnel/pull/536#discussion_r1238781334

### PR DESCRIPTION
**Pull request by nmittler**
Fixes https://github.com/bluejekyll/trust-dns/issues/1976.

### CODE REVIEW

**Comment by nmittler** [comment]
@djc @bluejekyll this is a WIP, but I wanted to send it out to get early feedback.

**Comment by nmittler** [comment]
@djc regarding your https://github.com/bluejekyll/trust-dns/issues/1976#issuecomment-1606108256, looks like we don't currently depend on hyper. Not sure if it would be acceptable to bring that in as a general dependency. Thoughts?

**Comment by djc** [comment]
I did not mean bringing in hyper as a dependency, merely copying the way it implements graceful shutdown (which seems like it might be lighter-weight than the linkerd stuff that you copied)?

**Comment by nmittler** [comment]
@djc I reworked this based on your comments regarding hyper. So far, I've only implemented UDP so we can see what it looks like.

**Comment by djc** [comment]
This seems to add a whole bunch of abstraction on top of what's there now just to add graceful shutdown. Instead, I think we should plug graceful shutdown functionality into the existing abstractions.

**Comment by nmittler** [comment]
@djc

> This seems to add a whole bunch of abstraction on top of what's there now just to add graceful shutdown. Instead, I think we should plug graceful shutdown functionality into the existing abstractions.

My first pass with the linkerd drain crate was just this. However, after looking into a more generic approach (as is done with hyper's Server API), I ran problems with the current structure of `ServerFuture`...

Ideally, when we start each server loop we would pass in a graceful shutdown future that can be used in `stream.take_until` to stop accepting incoming connections. However, if we pre-create the server loops when we `register` the socket, there's no easy way to insert a graceful shutdown future later on.

A few paths forward that come to mind:

1. Provide generic `registerXXX` methods.  This might be the least invasive approach. The idea here is that for every `register` method we have a generic version of the method that accepts a user-defined future for graceful shutdown. For example, we would have both ```fn register_socket(UdpSocket)``` and
```
pub fn register_socket_graceful<Fut, Out>(&mut self, socket: net::UdpSocket, drain: Fut)
    where
        Fut: Future<Output = Out> + Send + Unpin + 'static,
        Out: Send + Unpin + 'static {
    ...
    }
```
Note that because the future is internally being passed into `tokio::spawn`, both the `Future` and its `Output` have to be `static` to satisfy the lifetime requirements. 

2. Add the drain future as a member of `ServerFuture`. This actually gets tricky, since all of the generic parameters have to now be added to `ServerFuture`, itself. This makes it difficult to have a simple `new` method that takes only the handler, because there is no way to infer all of the other parameters.
3. Break apart the structure so that we don't create the server loops until later when the decision of whether to use graceful shutdown is made. In this case, the `registerXXX` methods would just add futures, and then we would have a `serve` or `run` method that takes an optional graceful shutdown future. This ends up requiring the most restructuring of the server code, since the storage of the futures requires a known size. I solved this in the new code by using a `Server` enum, and then having a `CompositeServer` that mainains a list of them.

I'm definitlely not a rust expert so there may better ways around some of the issues I was encountering. Although at the end of the day, the `Server` structure that I came up with seems pretty clean, albeit a bit of a departure from the current code.

I'm definitely open to suggestions.

@djc @bluejekyll thoughts?

**Comment by bluejekyll** [comment]
I'm coming to this conversation a little late. Looking briefly, I'm not sure why the new server is necessary?

Thinking about how I solve similar problems,I would probably just share an `AtomicBool` to all the handler loops, and then have the loops exit cleanly after all current requests complete? I haven't put a ton of thought into this right now, but is there a reason something as simple as that wouldn't work?

I haven't looked at the code in a lot of detail right now, but based on my memory of it I think this would be possible. Is this something you considered? (I haven't looked at the Hyper implementation mentioned). The Type would be a handle that would be returned by the server, something roughly like:

```rust
struct ServerHandle {
    is_shutdown: Arc<AtomicBool>,    // if false, then handlers should continue accepting work
}

impl ServerHandle {
    /// Tell the server handlers to shutdown
    pub fn shutdown(&self) {
        is_shutdown.store(true);
    }

    /// Handler loops would read this every loop iteration, stop if true
    pub fn is_shutdown(&self) -> bool {
        is_shutdown.load()
    }
}
```

The `ServerFuture` then already has this method to await tasks finishing, so it just becomes a task of figuring out how we ire the above through all the handlers: https://docs.rs/trust-dns-server/0.22.1/trust_dns_server/server/struct.ServerFuture.html#method.block_until_done

**Comment by nmittler** [comment]
@bluejekyll That's fair. The complexity currently is because I was attempting to allow an application to hook in their shutdown mechanism directly, similarly to what hyper does. This does introduce a lot of complexity. But I think you're right, we don't need to do that. Will update the PR shortly.

**Comment by bluejekyll** [comment]
@djc do you agree with this approach, would be good to get that feedback before @nmittler puts in a ton of work.

**Comment by nmittler** [comment]
@bluejekyll @djc I refactored things, PTAL when you get a moment.

File: crates/server/src/server/server_future.rs
Comment (nmittler) [inline crates/server/src/server/server_future.rs:742]: Note to reviewers: I think this always should have blocked until all tasks are completed. With the addition of graceful shutdown, this needed to be corrected.

File: crates/server/src/server/server_future.rs
Comment (nmittler) [inline crates/server/src/server/server_future.rs:731]: Alternatively, we could return both the future and a drain handle, to allow the application to handle them separately. I'm open to suggestions.

**Comment by nmittler** [comment]
Also, not sure why the build is failing to find `tokio::select!` ... everything builds fine for me locally.

**Comment by djc** [comment]
> @djc do you agree with this approach, would be good to get that feedback before @nmittler puts in a ton of work.

Yup, sounds like the right direction to me!

File: crates/server/src/server/https_handler.rs
Comment (djc) [inline crates/server/src/server/https_handler.rs:11]: Please keep the `std` imports in a separate block.

File: crates/server/src/server/quic_handler.rs
Comment (djc) [inline crates/server/src/server/quic_handler.rs:10]: Here, too.

File: crates/server/src/server/server_future.rs
Comment (djc) [inline crates/server/src/server/server_future.rs:773]: I think we should just skip this layer of abstraction, and just pass the `Arc<AtomicBool>` in directly. I also don't think the `Drain` naming is great in this context, I'd probably call it `shutting_down`.

File: crates/server/src/server/server_future.rs
Comment (djc) [inline crates/server/src/server/server_future.rs:813]: Why is this necessary?

File: crates/server/src/server/server_future.rs
Comment (djc) [inline crates/server/src/server/server_future.rs:21]: Please avoid making unrelated changes in the same commit/PR, this makes it harder to review the changes you've made.

File: crates/server/src/server/server_future.rs
Comment (djc) [inline crates/server/src/server/server_future.rs:68]: Why do we need this change?

File: crates/server/src/server/server_future.rs
Comment (djc) [inline crates/server/src/server/server_future.rs:733]: Why is this needed?

File: crates/server/src/server/server_future.rs
Comment (djc) [inline crates/server/src/server/server_future.rs:944]: Why is this test no longer valuable?

**Comment by djc** [comment]
> Also, not sure why the build is failing to find `tokio::select!` ... everything builds fine for me locally.

Probably because tokio's Cargo features aren't depended upon correctly.

File: crates/server/src/server/server_future.rs
Comment (nmittler) [reply crates/server/src/server/server_future.rs:68]: This just adds a line that wraps the original stream with one that will drain as we need.

File: crates/server/src/server/server_future.rs
Comment (nmittler) [reply crates/server/src/server/server_future.rs:733]: This allows the caller to use the API in a couple of different ways. They can either call `shutdown_gracefully`, which starts the shutdown and returns the shutdown future all in one shot. Alternatlively they can call `get_drain` and `block_until_done` separately, which allows the application to handle the two separately.

File: crates/server/src/server/server_future.rs
Comment (nmittler) [reply crates/server/src/server/server_future.rs:773]: This abstraction being a future allows us to use it in situtations like `take_until`, which is used for UDP. I also think the abstraction is pretty light, and doesn't complicate the API.

Regarding the naming, the term `drain` is a pretty standard term referring to the trigger that begins a graceful shutdown. Both hyper and linkerd mix `drain` and `graceful` throughout their API and docs. Calling `drain` is more concise than calling `start_graceful_shutdown`. I added a method `shutdown_gracefully` as an all-in-one helper that both starts the graceful shutdown and returns the waiting future.

File: crates/server/src/server/server_future.rs
Comment (nmittler) [reply crates/server/src/server/server_future.rs:813]: I moved where this is called to the outside of the loops, where there may or may not be tasks in the join_set. If `FutureExt::now_or_never` is called with None, it hangs forever.

File: crates/server/src/server/server_future.rs
Comment (nmittler) [reply crates/server/src/server/server_future.rs:944]: It is, I just renamed it to `abort` to distinguish it from the test for `graceful_shutdown`.

**Comment by nmittler** [comment]
@djc I resolved the issue with finding `tokio::select!` but now there is a different issue, that doesn't seem to be directly related to this PR: https://github.com/bluejekyll/trust-dns/actions/runs/5423528718/jobs/9861952221?pr=1977. Any thoughts?

**Comment by djc** [comment]
No, you'll need to dig in and figure out what's going on.

File: crates/server/src/server/server_future.rs
Comment (djc) [reply crates/server/src/server/server_future.rs:733]: I suppose my question is, why would someone want to handle the two separately?

**Comment by nmittler** [comment]
@bluejekyll @djc I've confirmed that there is a build issue on main that is unrelated to this PR: https://github.com/bluejekyll/trust-dns/pull/1982.

Not sure if something changed recently? Perhaps one of these?

https://github.com/bluejekyll/trust-dns/pull/1978
https://github.com/bluejekyll/trust-dns/pull/1979

**Comment by djc** [comment]
Cargo doesn't default to sticking to the locked dependencies. It's likely that the proc-macro2 crate is using some nightly feature that's newer than the nightly version we're using in CI or vice versa. Check its issue tracker?

**Comment by nmittler** [comment]
I understand that I have a vested interest in this getting fixed, but at the end of the day I'm just contributing here. This feels more like a task for a maintainer, although I suspect @bluejekyll may be the only one since it's his repo? I'll take a look when I get some cycles.

**Comment by nmittler** [comment]
FYI, looks like they're working on the issue: https://github.com/rust-lang/rust/issues/113152

Will see if I can throw together a temporary fix.

**Comment by nmittler** [comment]
@bluejekyll @djc I've converted https://github.com/bluejekyll/trust-dns/pull/1982 into a fix for rust nightly

File: crates/server/src/server/server_future.rs
Comment (nmittler) [reply crates/server/src/server/server_future.rs:733]: It's pretty common, actually. In fact, the Istio code does this for all of its server threads. If the application has its own shutdown hook (e.g. user clicks "x" on the application window, or some event is received on some channel), the shutdown hook can then trigger a drain of all the server threads. The code that triggers the shutdown would, in this case, be separate from the code that's waiting for the server to shutdown.

**Comment by nmittler** [comment]
@djc I think I've addressed your comments. Let me know if there is anything else.

**Comment by djc** [comment]
I'm mostly AFK for the next 2 weeks so am unlikely to be able to provide much feedback here.

**Comment by nmittler** [comment]
@bluejekyll @djc I ended up going back to using linkerd's `drain-rs` for controlling the graceful shutdown. The problem I was seeing with using `AtomicBool` as a `Future` is that it never triggers the `Waker`, which can lead to it shutdown hanging indefinitely. The `drain-rs` lib seems pretty light to me ... I don't think it should bloat the server much.

Additionally, this PR seems to be having issues with code coverage (currently at ~61%). I've updated all uses of `ServerFuture` that I found in our tests with the new graceful shutdown logic. Looking at the output, it appears that most of the missed coverage is for existing code. I suspect the original code landed before the code coverage analysis was added? Can we just force merge this PR as-is? Adding coverage to `ServerFuture` seems to be beyond the scope here.

**Comment by bluejekyll** [comment]
Ah, yes. I see the issue, needing to wake all the tasks on shutdown, I hadn't really been thinking about that, so thanks for pointing out the complexity here.

I guess the question I have, is drain-rs the simplest object type for this? I don't know much about that dependency, but in theory, any future aware channel would work just as well and avoid a new dependency, correct?

**Comment by nmittler** [comment]
@bluejekyll 

> in theory, any future aware channel would work just as well and avoid a new dependency, correct?

Are you thinking there is something we can leverage from existing dependencies here? Or are you suggesting that we roll our own? AFAICT the `drain` library [only depends on tokio](https://github.com/linkerd/drain-rs/blob/main/Cargo.toml#L17), and it's just using channels under the covers. I think it would be preferrable to just use that if the only alternative is to roll our own. WDYT?

**Comment by djc** [comment]
> @bluejekyll
> 
> > in theory, any future aware channel would work just as well and avoid a new dependency, correct?
> 
> Are you thinking there is something we can leverage from existing dependencies here? Or are you suggesting that we roll our own? AFAICT the `drain` library [only depends on tokio](https://github.com/linkerd/drain-rs/blob/main/Cargo.toml#L17), and it's just using channels under the covers. I think it would be preferrable to just use that if the only alternative is to roll our own. WDYT?

That sounds plausible to me.

**Comment by nmittler** [comment]
@bluejekyll @djc I've rebased and also added a small wrapper around the linkerd drain signal, so that we don't expose their API from `ServerFuture`. PTAL

Also, any thoughts on this?

> Additionally, this PR seems to be having issues with code coverage (currently at ~61%). I've updated all uses of `ServerFuture` that I found in our tests with the new graceful shutdown logic. Looking at the output, it appears that most of the missed coverage is for existing code. I suspect the original code landed before the code coverage analysis was added? Can we just force merge this PR as-is? Adding coverage to `ServerFuture` seems to be beyond the scope here.

**Comment by bluejekyll** [comment]
Thanks for all the explanations, @nmittler, I agree with this usage. I also just took a look at the code, seems reasonable enough to me. On the code coverage, I use that to really understand he risk we might have in regards to changes coming in, not something necessarily required for us to hit.

**Review by bluejekyll — APPROVED** [review]
Ok, I've reviewed. I think this overall looks good. I don't want to get pedantic on naming, but I'm not sure I would name the local variables `drain` but instead something more like `shutdown` or something that's a little closer to their usage.

@djc, can you close out some of your conversations if you don't have any open concerns?

Thanks for putting this all together @nmittler , I know this has taken a while, I appreciate you working through all the discussion points and taking the time to see this through.

**Comment by nmittler** [comment]
Thanks @bluejekyll. FYI, I just pushed a change to use `shutdown` instead of `drain`.

**Review by bluejekyll — APPROVED** [review]

**Comment by nmittler** [comment]
@djc any other comments/concerns before we land this?

File: crates/server/src/server/server_future.rs
Comment (bluejekyll) [reply crates/server/src/server/server_future.rs]: Thanks for the heads up, this looks good.

**Comment by bluejekyll** [comment]
Merging, thanks for getting this done, @nmittler!

**Comment by nmittler** [comment]
@bluejekyll what is the cadence for pushing new artifacts? I'd like to switch Istio over to using the graceful shutdown logic when it's available?

**Comment by bluejekyll** [comment]
Since we're on alpha's right now, I'd be happy to publish another release today.

**Comment by nmittler** [comment]
@bluejekyll excellent! Thanks for all your help!

**Comment by theduke** [comment]
@nmittler thanks for implementing this!

I unfortunately couldn't dedicate more time than creating the issue.

**Comment by caobug** [comment]
We shouldn't move ownership on shutdown_gracefully(), that complicates things.

**Comment by djc** [comment]
Please file a new issue or PR with suggestions on what we should do instead and what your use case is (why does this complicate things).

**Pull request merged**
